### PR TITLE
ENH: FreeSurfer Wrappings for Recon-All Workflow

### PR DESCRIPTION
**Pull request by dgellis90**
This pull request add the wrappings for the following FreeSurfer commands:
* mri_nu_correct.mni
* mri_watershed
* mri_normalize
* mri_ca_normalize
* mri_ca_register
* mri_ca_label
* mris_ca_label
* mri_cc
* mri_segment
* mri_edit_wm_with_aseg
* mri_concatenate_lta
* mri_label2label
* mris_label2annot
* mris_spherical_average
* mri_add_xform_to_header
* talairach_afd
* talairach_avi
* tal_QC_AZS
* mri_remove_neck
* mri_fill
* mris_inflate
* mris_sphere
* mris_fix_topology
* mris_euler_number
* mris_remove_intersection
* mris_make_surfaces
* mris_curvature
* mris_curvature_stats
* mris_jacobian
* mris_calc
* mris_volmask
* mris_anatomical_stats
* pctsurfcon
* mri_relabel_hypointensities
* mri_aparc2aseg
* apas2aseg
* mri_robust_template
* mri_fuse_segmentations
* mpr2mni305
* avi2talxfm
* mri_em_register
* mris_register
* mrisp_paint

This pull requests also adds inputs to some of the existing wrappings for FreeSurfer commands. These wrappings and changes were needed to create a a recon-all workflow in nipype that performs the same steps as recon-all. The workflows and usage information can be found here:
https://github.com/BRAINSia/BRAINSTools/tree/FreeSurferBreakout/AutoWorkup/fs_nipype

### CODE REVIEW

**Comment by satra** [comment]
thanks for this tremendous effort!

**Comment by satra** [comment]
it will take us a bit to review all the changes.

**Comment by chrisfilo** [comment]
Wow! This is very impressive!

**Comment by hjmjohnson** [comment]
@satra @chrisfilo We are preparing an OHBM abstract to describe this work.  It would be ideal to reference nipype proper rather than our branch for these changes (see very rough draft of abstract).

[Johnson_HBM_FreeSurferHPE.pdf](https://github.com/nipy/nipype/files/119921/Johnson_HBM_FreeSurferHPE.pdf)

Would you mind giving some initial gameplan for how we can move forward with getting these changes into the main branch?  Should we squash this into 1 patchset, or just a few patch sets?

Are there other obvious concerns that you have with this?

**Comment by satra** [comment]
@hjmjohnson - as soon as i saw this i was so excited i wrote to the freesurfer folks to get hold of their testing data. i will let you guys know when i see their test suite.

is there a place where you have testing results from running this and comparing different stages?

File: nipype/interfaces/freesurfer/longitudinal.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    input_spec = FuseSegmentationsInputSpec
+    output_spec = FuseSegmentationsOutputSpec
+
+    def _format_arg(self, name, spec, value):
+        if name in ('in_segmentations', 'in_segmentations_noCC', 'in_norms'):
+            # return enumeration value
+            return spec.argstr % os.path.basename(value[0])
+        return super(FuseSegmentations, self)._format_arg(name, spec, value)
+
+    def _list_outputs(self):
```
Comment (satra) [inline nipype/interfaces/freesurfer/longitudinal.py]: it seems name source and name template can take care of a few of these things. could you please go through and check where we can simplify the interface.

**Comment by satra** [comment]
also in addition to this, it would be great to add the freesurfer recon-all workflow into examples.

**Comment by hjmjohnson** [comment]
@dgellis90

David,  Please address all of Satra's e-mails above.  With guidance from the nipype team, we will make this a fabulous resource for further data processing with FreeSurfer or augmentations to freesurfer.

Hans

File: nipype/interfaces/freesurfer/longitudinal.py (hunk 1 above)
Comment (hjmjohnson) [inline nipype/interfaces/freesurfer/longitudinal.py]: @satra  I'm not quite sure of what you are asking for here.  Is there another nipype tools that we could use as a reference that uses "name source" and "name template" paradigm?

**Comment by hjmjohnson** [comment]
@satra We always intended to provide the recon-all workflows to the examples, but that was going to be a second patch set.  Do you think we should add that to this pull request?

**Comment by dgellis90** [comment]
@satra 
I ran the Mac OS X version of recon-all for a single subject and compared the results to the recon-all workflow that I wrote in nipype. The output images and surfaces were identical. You can see the results here:
[reconalloutputs.xlsx](https://github.com/nipy/nipype/files/121755/reconalloutputs.xlsx)

The output images were compared by converting them to NIFTI and then using SimpleITK to determine if they were identical. The surfaces were compared by converting them to VTK and using diff to determine if they were identical.

This testing was done back in August. I am currently working on rerunning the testing using the version of the FreeSurfer 6.0 beta from September.

**Comment by dgellis90** [comment]
@satra 

What do you mean by "name source and name template"? Could you provide an example?

Thanks,
David

**Comment by satra** [comment]
@dgellis90 - here you go:

dev docs: http://www.mit.edu/~satra/nipype-nightly/devel/cmd_interface_devel.html#creating-outputs-on-the-fly

http://www.mit.edu/~satra/nipype-nightly/devel/interface_specs.html#id2 (see the example)

for a simple example see (write before "workflow pieces"):

http://nbviewer.jupyter.org/urls/dl.dropbox.com/s/mxud29oj2mxid6o/nipype_dev_tutorial.ipynb

see the input spec of flirt, which does not have list outputs.

https://github.com/nipy/nipype/blob/a8956d2e26eb9b9d6b5df7f73b8de5c1a3191dcb/nipype/interfaces/fsl/preprocess.py#L522

**Comment by satra** [comment]
@dgellis90 - could you please merge with current master?

**Comment by hjmjohnson** [comment]
Satra,

I know that David is working hard to clean up and stay up-to-date with nipype.  We discussed this yesterday, and he hopes to have an update by this Friday.

Hans

--

**Comment by satra** [comment]
@hjmjohnson - thanks. this was mostly me checking in on some PRs. i'm going to spend a little more time on addressing things from now on :)

**Comment by hjmjohnson** [comment]
@Satra,  This is great. It is good to take PR’s to terminal state quickly.  Even if that means closing them as incomplete or needing further attention at some distant future.

We have found that with ITK if we keep the PR backlog small, community satisfaction goes up.

Hans

--

File: nipype/workflows/smri/freesurfer/utils.py [hunk 2]
```diff
@@ -371,3 +374,489 @@ def create_tessellation_flow(name='tessellate', out_format='stl'):
...
+                                                   'RB_all_2014-08-21.gca')
+    config['registration_template_withskull'] = os.path.join(config['fs_home'], 'average',
+                                                             'RB_all_withskull_2014-08-21.gca')
+    for hemi in ('lh', 'rh'):
+        config['{0}_atlas'.format(hemi)] = os.path.join(
+            config['fs_home'], 'average',
+            '{0}.average.curvature.filled.buckner40.tif'.format(hemi))
+        config['{0}_classifier'.format(hemi)] = os.path.join(
+            config['fs_home'], 'average',
+            'rh.curvature.buckner40.filled.desikan_killiany.2010-03-25.gcs'.format(hemi))
```
Comment (chrisfilo) [inline nipype/workflows/smri/freesurfer/utils.py]: Codacy Issue found: [Too many arguments for format string](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2302210766/issues/source?bid=2767207&fileBranchId=3084409#l451)

File: nipype/workflows/smri/freesurfer/autorecon3.py [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+
+        # WM/GM Contrast
+        contrast = pe.Node(Contrast(), name="WM_GM_Contrast_{0}".format(hemisphere))
+        contrast.inputs.hemisphere = hemisphere
+        contrast.inputs.copy_inputs = True
+
+        hemiwf2.connect([(hemi_inputspec2, contrast, [('orig_mgz', 'orig'),
+                                                      ('rawavg', 'rawavg'),
+                                                      ('{0}_white'.format(hemisphere), 'white'),
+                                                      ('cortex_label'.format(hemisphere), 'cortex'),
```
Comment (chrisfilo) [inline nipype/workflows/smri/freesurfer/autorecon3.py]: Codacy Issue found: [Too many arguments for format string](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2302210762/issues/source?bid=2767207&fileBranchId=3084409#l470)

**Comment by dgellis90** [comment]
@satra I added the freesurfer workflows and created an example (smri_fsreconall.py), but the cicleci tests are failing because during workflow setup, I require that the FREESURFER_HOME environmental variable be set. Should I just remove this requirement, or is there a way that I can get around this?

**Comment by dgellis90** [comment]
@satra I added name source to simplify a lot of the interfaces. I also added an example workflow that runs recon-all on two subjects from the tutorial data and then gets the average subject. Everything works when running the workflow using FreeSurfer 6.0 and the outputs look identical to the standard recon-all workflow.

**Comment by satra** [comment]
@dgellis90 - can you update the CHANGES file and push. i will merge it.

the freesurfer beta was yanked and they are planning to release 6.0 sometime before ohbm. have you tested this at all with freesurfer 5.3?

**Comment by dgellis90** [comment]
@satra I had been doing all my testing in 6.0 beta thinking it would be released soon. So I'll run some tests over the next few days and see if I can get it running on 5.3 as well.

**Comment by satra** [comment]
@dgellis90 - you will also need to merge with current master, you probably have an older version of the CHANGES file.

**Comment by satra** [comment]
@dgellis90 - let us know how tests with 5.3 go

**Comment by hjmjohnson** [comment]
@satra David and I discussed this today, and we hope that it will be completed yet this week.

**Pull request merged**